### ISSUE TITLE
Loss display maybe wrong for full batch

### ISSUE THREAD

**Issue by rperrot** [keras-team/keras#8578]
Hi keras devs,

first of all, thanks for your library, maybe the easiest way to get into deep learning.

I wanted to use keras to make some illustrations on SGD vs batch-SGD vs mini-batch SGD. I simulate batch SGD using mini batch size equal to the training set, but in that case, the error displayed after each epoch is shifted (ie: the error displayed after epoch 1 is the initial error, error of the second epoch is the error of the first epoch, ...) 

here is a simple test program on MNIST : 

```
import keras

from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense


(x_train, y_train), (x_test, y_test) = mnist.load_data()

img_rows = 28
img_cols = 28

x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)

# Map to [-1;1]
x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
x_train *= 2
x_test *= 2
x_train -= 1
x_test -= 1

y_train = keras.utils.to_categorical(y_train, 10)
y_test = keras.utils.to_categorical(y_test, 10)

model = Sequential()
model.add(Dense(10, activation='softmax', use_bias=False, kernel_initializer="zeros", input_shape=(784,)))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.SGD(),
              metrics=[keras.losses.categorical_crossentropy])


# Initial error (before training) ~ 2.3026
print( model.evaluate(x_train,y_train,batch_size=60000) )
model.fit(x_train,y_train,batch_size=60000,epochs=4,shuffle=False,verbose=2)
```

I got the following output : 

```
60000/60000 [==============================] - 0s 5us/step
[2.3026120662689209, 2.3026120662689209]
Epoch 1/4
 - 0s - loss: 2.3026 - categorical_crossentropy: 2.3026
Epoch 2/4
 - 0s - loss: 2.2567 - categorical_crossentropy: 2.2567
Epoch 3/4
 - 0s - loss: 2.2142 - categorical_crossentropy: 2.2142
Epoch 4/4
 - 0s - loss: 2.1733 - categorical_crossentropy: 2.1733
```

As you can see, evaluation gives the same result as the first epoch which is wrong, and if I ask only one epoch like that : 

```
print( model.evaluate(x_train,y_train,batch_size=60000) )
model.fit(x_train,y_train,batch_size=60000,epochs=1,shuffle=False,verbose=2)
print( model.evaluate(x_train,y_train,batch_size=60000) )
```

I got : 

```
60000/60000 [==============================] - 0s 3us/step
[2.3026120662689209, 2.3026120662689209]
Epoch 1/1
 - 0s - loss: 2.3026 - categorical_crossentropy: 2.3026
60000/60000 [==============================] - 0s 3us/step
[2.2566978931427002, 2.2566978931427002]
```

So the evaluation is not equal to the loss displayed after the epoch. 

Did I missed something, or the behavior is wanted ?

**Comment by ozabluda** [comment]
what happens if you don't call `evaluate` before training?

**Comment by rperrot** [comment]
It has no impact on the messages during training : 

```
Epoch 1/4
 - 0s - loss: 2.3026 - categorical_crossentropy: 2.3026
Epoch 2/4
 - 0s - loss: 2.2567 - categorical_crossentropy: 2.2567
Epoch 3/4
 - 0s - loss: 2.2142 - categorical_crossentropy: 2.2142
Epoch 4/4
 - 0s - loss: 2.1733 - categorical_crossentropy: 2.1733
```

**Comment by ozabluda** [comment]
This is intended behaviour. How would fit() know any different after epoch 1? It would have had to forward prop through the entire dataset again, which doesn't happen until epoch 2.

**Comment by rperrot** [comment]
I though the loss displayed is le loss after training the i-th epoch. So you mean it's the loss before the epoch ?

**Comment by ozabluda** [comment]
Loss is after epoch minus last minibatch. The only way to know the loss for the last minibatch is to forward prop through it again, which happens only in the next epoch.

**Comment by rperrot** [comment]
Ok, since I use minibatch equal to the size of the minibatch, the last minibatch is equal to the last epoch. It make sense, but it's not very intuitive. That means that the loss/metric does not corresponds to the loss of the epoch. 

Is there a link in the documentation that talk about this ? (I didn't read anything on when the metric is computed during fit() )

**Comment by ozabluda** [comment]
It even worse. The loss is average for the epoch, which changes
minibatch-to-minibatch.

**Comment by brge17** [comment]
@ozabluda 

Adding it to the validation won't change the fact that averaging mini-batches doesn't work for some metrics, and your training set is also now part of the validation set.

**Comment by geekbutton** [comment]
@brge17  

If the batch_size is smaller than the training set, then what is the loss mean for each mini-batch,and then what is the training loss.
In the documentation:
![image](https://user-images.githubusercontent.com/13257051/34257418-beee1e84-e694-11e7-8400-375fbfc93deb.png)
But it seems that the last minibatch loss is not the average loss over each batch of the training data. So , how can I get the training loss?
